libraries/Microsoft.Bot.Builder/AppInsightsLoggerMiddleware.cs

```diff
@@ -0,0 +1,151 @@
+        /// <param name="config"> (Optional) TelemetryConfiguration to use for Application Insights.</param>
+        public AppInsightsLoggerMiddleware(string instrumentationKey, bool logUserName = false, bool logOriginalMessage = false, TelemetryConfiguration config = null)
+        {
+            if (instrumentationKey == null)
+            {
+                throw new ArgumentNullException(nameof(instrumentationKey));
+            }
+
+            var telemetryConfiguration = config ?? new TelemetryConfiguration(instrumentationKey);
+            _telemetryClient = new TelemetryClient(telemetryConfiguration);
+            _logUserName = logUserName;
+            _logOriginalMessage = logOriginalMessage;
+        }
+
+        /// <summary>
+        /// Records incoming and outgoing activities to the Application Insights store.
+        /// </summary>
+        /// <param name="context">The context object for this turn.</param>
+        /// <param name="nextTurn">The delegate to call to continue the bot middleware pipeline.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        /// <seealso cref="ITurnContext"/>
+        /// <seealso cref="Bot.Schema.IActivity"/>
+        public async Task OnTurnAsync(ITurnContext context, NextDelegate nextTurn, CancellationToken cancellationToken)
+        {
+
+            context.Services.Add(AppInsightServiceKey, _telemetryClient);
+
+            // log incoming activity at beginning of turn
+            if (context.Activity != null)
+            {
+                if (string.IsNullOrEmpty((string)context.Activity.From.Properties["role"]))
+                {
+                    context.Activity.From.Properties["role"] = "user";
+                }
+
+                // Context properties for App Insights
+                if (!string.IsNullOrEmpty(context.Activity.Conversation.Id))
+                {
+                    _telemetryClient.Context.Session.Id = context.Activity.Conversation.Id;
+                }
+
+                if (!string.IsNullOrEmpty(context.Activity.From.Id))
+                {
+                    _telemetryClient.Context.User.Id = context.Activity.From.Id;
+                }
+
+                // Log the Application Insights Bot Message Received
+                try
+                {
+                    _telemetryClient.TrackEvent(BotMsgEvent, FillEventProperties(context));
+                }
+                catch (Exception ex)
+                {
+                    _telemetryClient.TrackException(ex);
+                }
+            }
+
+            // process bot logic
+            await nextTurn(cancellationToken).ConfigureAwait(false);
+        }
+
+        private static Dictionary<string, string> FillEventProperties(ITurnContext context)
```

Comment:
This code accesses the `context.Activity` property 10 times. I would suggest storing it a variable if for no other reason than reducing the noise of the code. Other benefits are reduced IL emitted, leading to smaller assemblies and eventually JIT overhead, never mind potential runtime overhead.

Similarly there are a couple places where the `.From`, `.Conversation` and other properties of the `Activity` are then accessed multiple times and those could benefit from some variables as well. It all adds up.

Reply:
Pulled out activity..